docs/device_resources.md

```diff
@@ -0,0 +1,224 @@
+- Logic BEL - A placable logic element.  May be subject to 0 or more placement constraints.
+- Site port BEL - A site port BEL represents a connection to a site pin contained within the parent tile of the site.  See [Site port BEL](#site_port_bel).
+- Routing BEL - A routing BEL connects at most 1 input BEL pin to the output BEL pin.  See [Routing BEL](#routing_bel).
+- Site PIP - A pair of input and output BEL pins belonging to a BEL that represents a logically connection.
+- Cell - A logical element of a design that some number of cell ports and some number of cell instances, and some number of nets.
+- Cell port - The boundry between the interior of a cell and the containing cell (if any).
+- Cell instance - A instance of a cell.  The cell ports of may be connected to nets within the parent cell.
+- Net - A set of logically connected cell ports.
+
+## The place and route problem
+
+The device resources schema is intended to provide a description for a tool
+solving the place and route problem. The definition of the problem used by
+this schema is described below:
+
+There exists exactly 1 **cell instance** (the **top** instance) that contains 1 or
+more leaf **cell instances** that must be placed at **BELs**, such that no
+**constraints** are violated and the **nets** between the **cell instances** are
+**routable**. **Routable** means that **site wires**, **site PIPs**,  **nodes**, **PIPs**
+can be assigned to at most 1 **net** such that each **net driver BEL pin** can
+reach each every **net sink BEL pin** on the **net**.
+
+The device resource format describes how **cell instances** can be legally
+placed at **BELs** and how **cell pins** relate to **BEL pins**.  When a
+**cell instance** is placed at a **BEL**, it may be subject to 0 or more
+**constraints**.
+
+**Nets** are divided into 3 categories.  A **signal** net represents a signal
+that is not either the constant logical 0 or constant logical 1 net.
+The constant logical 0 and constant logical 1 nets are special because they
+can have multiple drivers in the device description.  Routing resources that
+are always part of the constant logical 0 or constant logical 1 net are
+explicitly defined in the device resources scheam.  The constant logical 0 net
+is listed in the schema as the "gnd" type.  The constant logical 1 net is
+listed in the schema as the "vcc" type.
+
+### Rules for routing
+
+Fully routed signal nets always begin at a output/inout BEL pin, and always
+end at an input/inout BEL pin.  If a net enters a site, that net **must** end
+at an input/inout BEL pin.  It is not legal for a net to enter and leave a
+site.  If such a path is required, a pseudo PIP should be added to the schema.
+
+## Details
+
+### Net routing summary
+
+Each net start at the driver output/inout BEL pin.  The BEL pin will be
+connected to exactly 1 site wire.  If the net sink can be reached within the
+site, then the net can use site PIPs (through a routing BEL) to reach a site
```

Comment:
nitpick - site PIPs might be LUT route-throughs as well as routing BELs (aiui)

Reply:
Sure.  I've also folded this into the SLICE example.